torch/_dynamo/variables/builder.py

```diff
@@ -724,6 +729,15 @@ def index_source(key):
                 source=self.source,
                 guards=make_guards(GuardBuilder.ID_MATCH),
             )
+        elif (
+            issubclass(type(value), type)
+            and trace_rules.check(value) == TorchCtxManagerClassVariable
+        ):
+            return TorchCtxManagerClassVariable(
+                value,
+                source=self.source,
+                guards=make_guards(GuardBuilder.FUNCTION_MATCH),
+            )
```

Comment:
We could make things more generic with:
```suggestion
        elif trace_rules.check(value) is not None:
            return trace_rules.check(value).create_with_source(value, source=source)
```

Then define a create_with_source `@classmethod` on TorchCtxManagerClassVariable, that adds the guards and returns an instance.

That will allow us to simplify this long chain if `elif`s to just have one case for all the trace rules.